client/gutenberg/extensions/tiled-gallery/layouts.js

```diff
@@ -0,0 +1,131 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import find from 'lodash/find';
+import TokenList from '@wordpress/token-list';
+
+/**
+ * Internal dependencies
+ */
+import { LAYOUT_STYLES, MAX_COLUMNS, TILE_MARGIN } from './constants';
+
+const squareLayout = ( { columns, margin, width, tileCount } ) => {
+	columns = Math.min( MAX_COLUMNS, columns );
+	const tilesPerRow = columns > 1 ? columns : 1;
+	const marginSpace = tilesPerRow * margin * 2;
+	const size = Math.floor( ( width - marginSpace ) / tilesPerRow );
+	let remainderSize = size;
+	let tileSize = remainderSize;
+	const remainder = tileCount % tilesPerRow;
+	let remainderSpace = 0;
+
+	if ( remainder > 0 ) {
+		remainderSpace = remainder * margin * 2;
+		remainderSize = Math.floor( ( width - remainderSpace ) / remainder );
+	}
+
+	let c = 1;
+	let tilesInRow = 0;
+	const rows = [];
+	let row = {
+		tiles: [],
+	};
+
+	for ( let i = 0; i < tileCount; i++ ) {
+		if ( remainder > 0 && c <= remainder ) {
+			tileSize = remainderSize;
+		} else {
+			tileSize = size;
+		}
+
+		row.tiles.push( {
+			height: tileSize,
+			width: tileSize,
+		} );
+
+		c++;
+		tilesInRow++;
+
+		if ( tilesPerRow === tilesInRow || remainder + 1 === c ) {
+			rows.push( row );
+			tilesInRow = 0;
+
+			row.height = tileSize + margin * 2;
+			row.width = width;
+			row.groupSize = tileSize + 2 * margin;
+
+			row = {
+				tiles: [],
+			};
+		}
+	}
+
+	if ( row.tiles.length > 0 ) {
+		row.height = tileSize + margin * 2;
+		row.width = width;
+		row.groupSize = tileSize + 2 * margin;
+
+		rows.push( row );
+	}
+
+	return rows;
+};
+
+// @TODO
+const rectangularLayout = options => squareLayout( options );
+const columnsLayout = options => squareLayout( options );
```

Comment:
Note how columns and rectangular (="mosaic") simply implement square layouts for now.